cpp/oneapi/dal/detail/communicator.hpp

```diff
@@ -0,0 +1,741 @@
+
+#ifdef ONEDAL_DATA_PARALLEL
+    template <typename T, enable_if_primitive_t<T>* = nullptr>
+    spmd_request allgather(sycl::queue& q,
+                           const T* send_buf,
+                           std::int64_t send_count,
+                           T* recv_buf,
+                           std::int64_t recv_count) const {
+        return allgather(q,
+                         reinterpret_cast<const byte_t*>(send_buf),
+                         send_count,
+                         reinterpret_cast<byte_t*>(recv_buf),
+                         recv_count,
+                         make_data_type<T>());
+    }
+#endif
+
+    template <typename T, enable_if_primitive_t<T>* = nullptr>
+    spmd_request allgather(const array<T>& send, const array<T>& recv) {
+#ifdef ONEDAL_ENABLE_ASSERT
+        check_if_same_send_count(send.get_count(), get_default_root_rank());
+
+        // Check if recv allocated count is enough to store all the received values
+        const std::int64_t min_recv_count = get_min_recv_count(send.get_count());
+        ONEDAL_ASSERT(recv.get_count() >= min_recv_count);
+#endif
+
+        if (send.get_count() == 0) {
+            ONEDAL_ASSERT(recv.get_count() == 0);
+            return spmd_request{};
+        }
+
+        ONEDAL_ASSERT(send.get_count() > 0);
+        ONEDAL_ASSERT(recv.has_mutable_data());
+
+        spmd_request request;
+
+        __ONEDAL_IF_QUEUE__(send.get_queue(), {
+            auto q = send.get_queue().value();
+
+            ONEDAL_ASSERT(recv.get_queue().has_value());
+            ONEDAL_ASSERT(recv.get_queue().value() == q);
+
+            request = allgather(q,
+                                send.get_data(),
+                                send.get_count(),
+                                recv.get_mutable_data(),
+                                send.get_count());
+        });
+
+        __ONEDAL_IF_NO_QUEUE__(send.get_queue(), {
+            request = allgather(send.get_data(),
+                                send.get_count(),
+                                recv.get_mutable_data(),
+                                send.get_count());
+        });
+
+        return request;
+    }
+
+    /// Combines data from all ranks using reduction operation and
+    /// distributes the result back to all ranks
+    ///
+    /// @param send_buf The send buffer
+    /// @param recv_buf The receiving buffer
+    /// @param count    The number of elements of `dtype` sent to and
+    ///                 received from each rank
+    /// @param dtype    The type of elements in the passed buffers
+    /// @param op       The reduction operation
+    ///
+    /// @return The object to track the progress of the operation
+    spmd_request allreduce(const byte_t* send_buf,
+                           byte_t* recv_buf,
+                           std::int64_t count,
+                           const data_type& dtype,
+                           const spmd_reduce_op& op = spmd_reduce_op::sum) const;
```

Comment:
Is it necessary to have a default value here? Explicit value looks more informative.

Reply:
It's matter of convention. To me having summation as default behavior is reasonable -- in most algorithms summation is required.